### ISSUE TITLE
java.io.EOFException when running tests

### ISSUE THREAD

**Issue by chipsenkbeil** [scalatest/scalatest#770]
Using `3.0.0-M11` of scalatest (need at least... 3.0.0-M9, I think, to get around a parallel test issue).

My tests used to run fine, but I introduced this problem at some point in my current branch. At some point, the tests fail with this exception (sporadically). Any idea what could trigger this? I'm going to have to revert a fairly-lengthy branch if I cannot figure this out.

```scala
Exception in thread "Thread-1" Exception in thread "Thread-5" java.io.EOFException
	at java.io.ObjectInputStream$BlockDataInputStream.peekByte(ObjectInputStream.java:2598)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1318)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:370)
	at sbt.React.react(ForkTests.scala:114)
	at sbt.ForkTests$$anonfun$mainTestTask$1$Acceptor$2$.run(ForkTests.scala:74)
	at java.lang.Thread.run(Thread.java:745)
java.io.EOFException
	at java.io.ObjectInputStream$BlockDataInputStream.peekByte(ObjectInputStream.java:2598)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1318)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:370)
	at org.scalatest.tools.Framework$ScalaTestRunner$Skeleton$1$React.react(Framework.scala:813)
	at org.scalatest.tools.Framework$ScalaTestRunner$Skeleton$1.run(Framework.scala:802)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by bvenners** [comment]
Hmm, this looks like a deserialization problem. Not much info there to go on. When this happens do you havce any other problem, like tests aborting, etc., or is it just noise in otherwise OK-looking results?

**Comment by chipsenkbeil** [comment]
Looks like it just stops running tests. For the project, there are currently 840 unit tests of which only 458 are run in the picture below.

<img width="1042" alt="screen shot 2015-11-10 at 1 58 57 pm" src="https://cloud.githubusercontent.com/assets/2481802/11074064/3b7db182-87b3-11e5-8e85-95c725cd248b.png">

You can see the builds here: https://travis-ci.org/chipsenkbeil/scala-debugger-api/builds/90378966

Think I found the commit that introduced the failure, but don't know what code is causing the issue.

**Comment by bvenners** [comment]
Can you try running it not in forked mode and see if you get a failure? Perhaps the root cause is getting masked by the deserialization problem (which we also want to fix).

**Comment by chipsenkbeil** [comment]
Oh, and these errors are only happening on Travis CI (in Docker container), not on my laptop. Going to push a commit with no forking of unit tests. This may or may not work as I had to fork before to avoid linkage issues against the JDK `tools.jar`.

**Comment by chipsenkbeil** [comment]
One last update. Ran with `-U` (unformatted mode) to see if I could get any additional information. Doesn't look like it fails consistently at all with regard to which spec or test. Tomorrow, I will try reverting back to the last commit before this specific failure and make sure that I don't encounter the problem.

**Comment by chipsenkbeil** [comment]
FYI, I still run into this issue for my integration tests occasionally. Just hit it here: https://travis-ci.org/chipsenkbeil/scala-debugger-api/jobs/97065319

I'm not able to run them in parallel yet, which is why I haven't been able to use that trick to avoid this problem.

**Comment by chipsenkbeil** [comment]
Seen it again with latest build under M14. https://travis-ci.org/ensime/scala-debugger/jobs/112349909#L1738

**Comment by melodious** [comment]
I see this problem regularly. We have tests in various projects running in parallel but not in parallel within a given project.

**Comment by chipsenkbeil** [comment]
Gotten worse again now that I've added more integration tests. All six of my Travis builds for integration tests failed with this issue.

https://travis-ci.org/ensime/scala-debugger/builds/115750923

**Comment by chipsenkbeil** [comment]
Also, somehow got parallel test execution to work on my integration tests, but that doesn't seem to fix this issue anymore. So, really hope to find some sort of solution soon.

**Comment by cheeseng** [comment]
@chipsenkbeil @bvenners Unfortunately it is still not clear to us what is causing the serialization problem, I think the line that blows up is this in 3.0.x branch: 

https://github.com/scalatest/scalatest/blob/3.0.x/scalatest/src/main/scala/org/scalatest/tools/Framework.scala#L790

It is the thread at the main process receiving events sent back from the forked processes, following the current code if it blows up the thread will terminate and no more events from forked processes will be reported further, which I think at least this behaviour concur with what you're observing.

I wonder if adding a try and catch there to print out full stack trace would give us more clue about what's going wrong?

**Comment by cheeseng** [comment]
@chipsenkbeil @bvenners FYI here could be a good place to add in a e.printStackTrace(), I think: 

https://github.com/scalatest/scalatest/blob/3.0.x/scalatest/src/main/scala/org/scalatest/tools/Framework.scala#L781

**Comment by chipsenkbeil** [comment]
I'll give that a shot and let you know what I find.

**Comment by chipsenkbeil** [comment]
Actually, I  just realized that this isn't occurring locally for me. Just on Travis. So, going to have to figure out a way to get this kind of change bubbled up to Travis.

**Comment by chipsenkbeil** [comment]
FYI, switched to a Drone cluster and have not hit the issue, although I'm only running integration tests twice (once for Scala 2.10.5 and once for Scala 2.11.6) instead of six times. So, maybe something in Travis' environment is more conducive to this issue?

**Comment by cheeseng** [comment]
@chipsenkbeil fyi I just deployed a 3.0.0-M16-SNAP1 that includes the printStackTrace() based on current 3.0.x branch, hopefully that can give us more info what's wrong on travis.  :)

**Comment by chipsenkbeil** [comment]
Awesome. I'll make a branch of my current codebase that uses that version and re-enables Travis to see what is going on. Will get to that today or tomrrow since you did the hard part for me. :)

**Comment by chipsenkbeil** [comment]
https://travis-ci.org/ensime/scala-debugger/builds/116037459

Here is a build using that snapshot. I don't see anything new in terms of stacktrace, though. You'll see the error in the integration tests. Unit tests are failing because of some incompatibility with ScalaMock, which is unrelated.

https://travis-ci.org/ensime/scala-debugger/jobs/116037461#L4392

**Comment by cheeseng** [comment]
@chipsenkbeil FYI we are digging to see what's wrong, on the way we fixed a serialization bug in WebBrowser, but I think it is probably a different issue with yours.  I wonder if it is possible that the forked process crashed when runs on travis.  Tomorrow I think I could try to use your 161 branch to submit another PR, that uses another scalatest version that includes more debugging info.

**Comment by cheeseng** [comment]
@chipsenkbeil @bvenners FYI, found this and I think it is probably related: 

https://github.com/sbt/sbt/issues/653

I'll try the memory fix.

**Comment by cheeseng** [comment]
@chipsenkbeil @bvenners Setting sudo = true seems to fix the problem, it is still hard to tell what's exactly wrong, since it is integration test, do you think it is possible that the tests created too many sub-processes that exhaust the system memory, causing the OS to start killing processes?

**Comment by cheeseng** [comment]
See my testing PR here: 

https://github.com/ensime/scala-debugger/pull/162

**Comment by chipsenkbeil** [comment]
@cheeseng @bvenners  that could definitely be the case, as I added an option to run the integration tests in parallel. Previously, they were running sequentially and hit the same issue, though. Furthermore, I encountered the same error in my unit tests when I did _not_ run them in parallel. There's more detail about that earlier in this issue.

I could try to whip up a branch that has the unit tests failing for the same reason if that would help.

**Comment by cheeseng** [comment]
@chipsenkbeil @bvenners Sure, one thing we can do is just change sudo := true and see if it fixes it, travis's max memory is 3G, while it might not be easy to measure, it would be good to cap the memory for the main process + the forked process to be < 3G at any point of time.  That would mean: 

main process + SBT forked process + debugger process in test (if you created)

should not be > 3G

**Comment by chipsenkbeil** [comment]
Okay, ran with fork memory improvements and running with `sudo: true` in the Travis yaml file. Looks like it resolved the exception, although two of the integration test runs encountered a buffered overflow, not that I know if it is related to ScalaTest or not.

https://travis-ci.org/ensime/scala-debugger/builds/118623961
https://travis-ci.org/ensime/scala-debugger/jobs/118623973 <-- buffered overflow

I still haven't fixed an issue after rebasing, so the language/repl modules don't compile, but the main module (scala-debugger-api) compiles and runs, which was the one that was encountering this issue. So, if the buffered overflow problem is unrelated, providing the memory caps on the main process, forked, process, and children worked.

**Comment by cheeseng** [comment]
@chipsenkbeil @bvenners the following issues seems to be related to the buffer overflow problem that you're seeing: 

https://github.com/travis-ci/travis-ci/issues/5227

**Comment by jeffrey-aguilera** [comment]
I have the same problem (exact same stack trace) using Scalatest 2.2.6 on Bamboo.  When this occurs, the tests abort with a failure.

**Comment by jeffrey-aguilera** [comment]
Upgraded to 3.0.0-RC4 ... same problem:
```
error	09-Jul-2016 01:10:17	Exception in thread "Thread-39" Exception in thread "Thread-35" java.io.EOFException
error	09-Jul-2016 01:10:17		at java.io.ObjectInputStream$BlockDataInputStream.peekByte(ObjectInputStream.java:2598)
error	09-Jul-2016 01:10:17		at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1318)
error	09-Jul-2016 01:10:17		at java.io.ObjectInputStream.readObject(ObjectInputStream.java:370)
error	09-Jul-2016 01:10:17		at org.scalatest.tools.Framework$ScalaTestRunner$Skeleton$1$React.react(Framework.scala:798)
error	09-Jul-2016 01:10:17		at org.scalatest.tools.Framework$ScalaTestRunner$Skeleton$1.run(Framework.scala:787)
error	09-Jul-2016 01:10:17		at java.lang.Thread.run(Thread.java:745)
error	09-Jul-2016 01:10:17	java.io.EOFException
error	09-Jul-2016 01:10:17		at java.io.ObjectInputStream$BlockDataInputStream.peekByte(ObjectInputStream.java:2598)
error	09-Jul-2016 01:10:17		at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1318)
error	09-Jul-2016 01:10:17		at java.io.ObjectInputStream.readObject(ObjectInputStream.java:370)
error	09-Jul-2016 01:10:17		at sbt.React.react(ForkTests.scala:114)
error	09-Jul-2016 01:10:17		at sbt.ForkTests$$anonfun$mainTestTask$1$Acceptor$2$.run(ForkTests.scala:74)
error	09-Jul-2016 01:10:17		at java.lang.Thread.run(Thread.java:745)
```

**Comment by mcamou** [comment]
this just started happening to me on 3.0.3

**Comment by lloydmeta** [comment]
Ran into this one for the first time [here](https://travis-ci.org/m3dev/octoparts/builds/249504075#L1704), also on Travis.

**Comment by holdenk** [comment]
So this appears to be either masking an OOM (either JVM or native I'm not sure). Changing sudo could impact this since sudo to required or true puts your process on a larger VM and decreasing the amount of memory in the fork (or parent sbt process) will also leave more memory free. I wish this was better communicated but at least its an easy fix once you find the issue.

**Comment by master-elodin** [comment]
I had the same issue in my tests. The issue turned out to be a test that was calling `System.setProperty("user.dir", "some value")`.

SBT never showed me an error related to that, but wrapping that call in another method that was then mocked fixed the issue for me. Not sure why it's silently failing on one error but throwing a different error, though.

**Comment by skestle** [comment]
Well, my particular issue stemmed from making stable defaults for deploys.
`akka.coordinated-shutdown.exit-jvm = on` is awesome for basic systems deployed as linux services, but it wrecks havoc in actor tests.

Turning it `off` in the test reference.conf restores normal operation.

(I don't claim this helps fix the issue, but while it remains open I'd expect many would like to have ideas on where to look to resolve their issues)
